sycl/source/detail/device_impl.cpp

```diff
@@ -22,17 +22,19 @@ namespace detail {
 /// Constructs a SYCL device instance using the provided
 /// UR device instance.
 device_impl::device_impl(ur_device_handle_t Device, platform_impl &Platform,
-                         device_impl::private_tag)
+                         device_impl::private_tag, bool IsSubDevice)
     : MDevice(Device), MPlatform(Platform),
       // No need to set MRootDevice when MAlwaysRootDevice is true
       MRootDevice(Platform.MAlwaysRootDevice
                       ? nullptr
                       : get_info_impl<UR_DEVICE_INFO_PARENT_DEVICE>()),
       // TODO catch an exception and put it to list of asynchronous exceptions:
       MCache{*this} {
-  // Interoperability Constructor already calls DeviceRetain in
-  // urDeviceCreateWithNativeHandle.
-  getAdapter().call<UrApiKind::urDeviceRetain>(MDevice);
+  if (!IsSubDevice) {
+    // Interoperability Constructor already calls DeviceRetain in
+    // urDeviceCreateWithNativeHandle.
+    getAdapter().call<UrApiKind::urDeviceRetain>(MDevice);
+  }
```

Comment:
The destructor `device_impl::~device_impl` calls `urDeviceRelease`, which is meant to balance the `urDeviceRetain` in the constructor. However, after this change, `urDeviceRetain` is no longer called for subdevices in the constructor, while `urDeviceRelease` is still invoked for them in the destructor. This imbalance is somewhat concerning. It seems that the actual root cause of the memory leak may lie elsewhere. Could you please clarify what you believe the underlying issue is?

Reply:
I am testing not calling `urDeviceRetain()` in `device_impl` ctor at all (https://github.com/intel/llvm/pull/20144) - one unittest will have to be changed at least ...